### ISSUE TITLE
Compute shader mem move inside `if` statement not working for different work group sizes on  GL backend

### ISSUE THREAD

**Issue by Stock84-dev** [gfx-rs/naga#1434]
**Description**
Compute shader algorithm: Initialize buffer with `local_invocation_id.x` then move the right half on top of left half.
I've removed `workgroupBarrier();` because they aren't the issue.
This also happens on arrays with `workgroup` storage class.
Results with one work group:
workgroup_size(2)   [1, 1] // correct
workgroup_size(4)   [2, 3, 2, 3] // correct
workgroup_size(8)   [4, 5, 6, 7, 4, 5, 6, 7] // correct
workgroup_size(16) [0, 0, 0, 0, 0, 0, 0, 0, 8, 9, 10, 11, 12, 13, 14, 15]
workgroup_size(32) [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31]

**Repro steps**

<details>
  <summary>Cargo.toml</summary>

  ```toml
 [package]
name = "demo"
version = "0.1.0"
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
env_logger = "0.9"
wgpu = "0.10"
pollster = "0.2"
bytemuck = { version = "1.4", features = [ "derive" ] }
anyhow = "1.0"
  ```
</details>

<details>

  <summary>main.rs</summary>

```rs
use std::sync::Arc;
use std::thread;
use wgpu::{BufferAddress, Maintain, MapMode};
macro_rules! src {
    () => {
        "\
[[block]]
struct Data {{
    data: array<u32>;
}};
[[group(0), binding(0)]] var<storage, read_write> dest : Data;

[[stage(compute), workgroup_size({size})]]
fn main(
    [[builtin(local_invocation_id)]] local_invocation_id: vec3<u32>,
) {{
    let lid = local_invocation_id.x;
    dest.data[lid] = lid;
    workgroupBarrier();
    if (lid < {half}u) {{
        dest.data[lid] = dest.data[lid + {half}u];
        // dest.data[lid] = dest.data[lid] * 2u; // this line would work fine
    }}
}}
 "
    };
}

pub fn main() -> anyhow::Result<()> {
    env_logger::init();
    pollster::block_on(bug())?;
    Ok(())
}

pub async fn bug() -> anyhow::Result<()> {
    let instance = wgpu::Instance::new(wgpu::Backends::all());
    let adapter = instance
        .request_adapter(&wgpu::RequestAdapterOptions {
            power_preference: wgpu::PowerPreference::default(),
            compatible_surface: None,
        })
        .await
        .unwrap();
    let (device, queue) = adapter
        .request_device(
            &wgpu::DeviceDescriptor {
                features: wgpu::Features::empty(),
                limits: wgpu::Limits::default(),
                label: None,
            },
            None,
        )
        .await
        .unwrap();
    let good_size = 8;
    let bad_size = 16;
    let good_src = format!(src!(), size = good_size, half = good_size / 2);
    let bad_src = format!(src!(), size = bad_size, half = bad_size / 2);
    println!("{}", good_src);
    println!("{}", bad_src);

    println!("{:#?}", adapter.get_info());
    let device = Arc::new(device);
    let dest_buffer = device.create_buffer(&wgpu::BufferDescriptor {
        label: Some("Vertex Buffer"),
        size: (std::mem::size_of::<f32>() * bad_size) as _,
        usage: wgpu::BufferUsages::STORAGE | wgpu::BufferUsages::MAP_READ,
        mapped_at_creation: false,
    });
    let d = device.clone();
    thread::spawn(move || loop {
        d.poll(Maintain::Wait);
    });

    let good_shader = device.create_shader_module(&wgpu::ShaderModuleDescriptor {
        label: Some("Shader"),
        source: wgpu::ShaderSource::Wgsl(good_src.into()),
    });
    let bad_shader = device.create_shader_module(&wgpu::ShaderModuleDescriptor {
        label: Some("Shader"),
        source: wgpu::ShaderSource::Wgsl(bad_src.into()),
    });

    let compute_bind_group_layout =
        device.create_bind_group_layout(&wgpu::BindGroupLayoutDescriptor {
            entries: &[wgpu::BindGroupLayoutEntry {
                binding: 0,
                visibility: wgpu::ShaderStages::COMPUTE,
                ty: wgpu::BindingType::Buffer {
                    ty: wgpu::BufferBindingType::Storage { read_only: false },
                    has_dynamic_offset: false,
                    min_binding_size: None,
                },
                count: None,
            }],
            label: None,
        });
    let compute_pipeline_layout = device.create_pipeline_layout(&wgpu::PipelineLayoutDescriptor {
        label: Some("compute"),
        bind_group_layouts: &[&compute_bind_group_layout],
        push_constant_ranges: &[],
    });

    let good_pipeline = device.create_compute_pipeline(&wgpu::ComputePipelineDescriptor {
        label: Some("Compute pipeline"),
        layout: Some(&compute_pipeline_layout),
        module: &good_shader,
        entry_point: "main",
    });
    let bad_pipeline = device.create_compute_pipeline(&wgpu::ComputePipelineDescriptor {
        label: Some("Compute pipeline"),
        layout: Some(&compute_pipeline_layout),
        module: &bad_shader,
        entry_point: "main",
    });

    let bind_group = device.create_bind_group(&wgpu::BindGroupDescriptor {
        layout: &compute_bind_group_layout,
        entries: &[wgpu::BindGroupEntry {
            binding: 0,
            resource: dest_buffer.as_entire_binding(),
        }],
        label: None,
    });

    let mut command_encoder =
        device.create_command_encoder(&wgpu::CommandEncoderDescriptor { label: None });
    let mut cpass =
        command_encoder.begin_compute_pass(&wgpu::ComputePassDescriptor { label: None });
    cpass.set_pipeline(&good_pipeline);
    cpass.set_bind_group(0, &bind_group, &[]);
    cpass.dispatch(1 as _, 1, 1);
    drop(cpass);
    queue.submit(Some(command_encoder.finish()));
    let slice = dest_buffer.slice(..(good_size / 2 * std::mem::size_of::<u32>()) as BufferAddress);
    slice.map_async(MapMode::Read).await?;
    let map = slice.get_mapped_range();
    let good_output: Vec<u32> = bytemuck::cast_slice(map.as_ref()).to_vec();
    drop(map);
    dest_buffer.unmap();

    let mut command_encoder =
        device.create_command_encoder(&wgpu::CommandEncoderDescriptor { label: None });
    let mut cpass =
        command_encoder.begin_compute_pass(&wgpu::ComputePassDescriptor { label: None });
    cpass.set_pipeline(&bad_pipeline);
    cpass.set_bind_group(0, &bind_group, &[]);
    cpass.dispatch(1 as _, 1, 1);
    drop(cpass);
    queue.submit(Some(command_encoder.finish()));
    let slice = dest_buffer.slice(..(bad_size / 2 * std::mem::size_of::<u32>()) as BufferAddress);
    slice.map_async(MapMode::Read).await?;
    let map = slice.get_mapped_range();
    let bad_output: Vec<u32> = bytemuck::cast_slice(map.as_ref()).to_vec();

    let good_values: Vec<_> = ((good_size / 2)..good_size)
        .into_iter()
        .map(|x| x as u32)
        .collect();
    let bad_values: Vec<_> = ((bad_size / 2)..bad_size)
        .into_iter()
        .map(|x| x as u32)
        .collect();
    assert_eq!(good_values, good_output);
    println!("good passed");
    assert_eq!(bad_values, bad_output);
    Ok(())
}

  ```
</details>

Doesn't work with this adapter:
```
AdapterInfo {
    name: "llvmpipe (LLVM 12.0.1, 256 bits)",
    vendor: 0,
    device: 0,
    device_type: Cpu,
    backend: Gl,
}
```

~~Would work when #1428 gets fixed.~~
Edit:
This adapter is always wrong
workgroup_size(2) [0, 0, 0, 0, 1]
workgroup_size(4) [0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 3]
workgroup_size(8)  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, 0, 0, 0, 5, 0, 0, 0, 6, 0, 0, 0, 7]
```
AdapterInfo {
    name: "AMD TURKS (DRM 2.50.0 / 5.14.2-1-MANJARO, LLVM 12.0.1)",
    vendor: 0,
    device: 0,
    device_type: DiscreteGpu,
    backend: Gl,
}
```

**Comment by Stock84-dev** [comment]
**Input**
```rust

[[block]]
struct Data {
    data: array<u32>;
};
[[group(0), binding(0)]] var<storage, read_write> dest : Data;

[[stage(compute), workgroup_size(16)]]
fn bad(
    [[builtin(local_invocation_id)]] local_invocation_id: vec3<u32>,
) {
    let lid = local_invocation_id.x;
    dest.data[lid] = lid;
    workgroupBarrier();
    if (lid < 8u) {
        dest.data[lid] = dest.data[lid + 8u];
        // dest.data[lid] = dest.data[lid] * 2u; // this line would work fine
    }
}
```

**Generated glsl**
```glsl
#version 310 es

precision highp float;
precision highp int;

layout(local_size_x = 16, local_size_y = 1, local_size_z = 1) in;

buffer Data_block_0Cs {
    uint data[];
} _group_0_binding_0;


void main() {
    uvec3 local_invocation_id = gl_LocalInvocationID;
    uint lid = local_invocation_id.x;
    _group_0_binding_0.data[lid] = lid;
    groupMemoryBarrier();
    if ((lid < 8u)) {
        uint _e13 = _group_0_binding_0.data[(lid + 8u)];
        _group_0_binding_0.data[lid] = _e13;
        return;
    } else {
        return;
    }
}
```

**Comment by kvark** [comment]
What does the shader look like with #1428?

**Comment by Stock84-dev** [comment]
workgroup_size(2) [0, 0, 0, 0, 1]
workgroup_size(4) [0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 3]
workgroup_size(8) [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, 0, 0, 0, 5, 0, 0, 0, 6, 0, 0, 0, 7]

**Comment by kvark** [comment]
I mean the shader itself

**Comment by Stock84-dev** [comment]
Whoops
It is identical as above. I guess that `llvmpipe` doesn't use 16 byte alignment by default.

**Comment by kvark** [comment]
I'm confused. How can it be "identical as above" if the PR adds `std430` to the structures?

**Comment by Stock84-dev** [comment]
Ah, I only changed version specification in Cargo.toml of wgpu not realizing that it was optional and that other packages require naga. 
When compiling wgpu i get this error.
```
Compiling naga v0.6.0 (https://github.com/gfx-rs/naga?rev=a4bf25a0d5b87a7f0fdd85eac1e2e5041d536f21#a4bf25a0)
   Compiling wgpu-hal v0.10.1 (/tmp/wgpu/wgpu-hal)
error[E0308]: mismatched types
   --> /tmp/wgpu/wgpu-core/src/validation.rs:909:78
    |
909 |                 Self::populate(&mut ep.inputs, arg.binding.as_ref(), arg.ty, &module.types);
    |                                                                              ^^^^^^^^^^^^^ expected struct `Arena`, found struct `UniqueArena`
    |
    = note: expected reference `&Arena<naga::Type>`
               found reference `&UniqueArena<naga::Type>`

error[E0308]: mismatched types
   --> /tmp/wgpu/wgpu-core/src/validation.rs:916:21
    |
916 |                     &module.types,
    |                     ^^^^^^^^^^^^^ expected struct `Arena`, found struct `UniqueArena`
    |
    = note: expected reference `&Arena<naga::Type>`
               found reference `&UniqueArena<naga::Type>`
```

It does generate correct GLSL now:
```glsl
#version 310 es

precision highp float;
precision highp int;

layout(local_size_x = 16, local_size_y = 1, local_size_z = 1) in;

layout(std430) buffer Data_block_0Cs {
    uint data[];
} _group_0_binding_0;


void main() {
    uvec3 local_invocation_id = gl_LocalInvocationID;
    uint lid = local_invocation_id.x;
    _group_0_binding_0.data[lid] = lid;
    groupMemoryBarrier();
    if ((lid < 8u)) {
        uint _e13 = _group_0_binding_0.data[(lid + 8u)];
        _group_0_binding_0.data[lid] = _e13;
        return;
    } else {
        return;
    }
}

```

**Comment by kvark** [comment]
Ok, so we need to either back-port #1435 to crates, or release the new Naga version

**Comment by teoxoy** [comment]
Closing since this has been fixed and release 0.7 includes the fix.
